### PR TITLE
Handle inconsistencies in Pantheon API response: for ddev issue #435

### PR DESCRIPTION
**Pull request by nmccrory**
## The Problem:
The Pantheon API is inconsistent in the datatypes they use for certain fields. We specifically noticed this in a field named `Verify` in type `Site`, that used to be a long integer and is now a string. Trying to convert that string to jsonInt64 would present the user with `invalid character '9' in literal false (expecting 'a')`. 
## The Fix:
Omit `Verify` and other unnecessary fields in pkg/go-pantheon's `Site` struct. 
## The Test:
Using a mock Pantheon API response, that is known to cause chaos, convert the response to a native Go type (in our case `Site`). 

To make the test fail and expose the bug through go-pantheon, add this field to `Site` in `site.go`:
```
Verify jsonInt64 `json:"verify"`
``` 
## Automation Overview:
Added `TestSite()` to `site_test.go`.
## Related Issue Link(s):
ddev issue [#435](https://github.com/drud/ddev/issues/435)

### CODE REVIEW

**Comment by rfay** [comment]
I changed the title. If it's "fixes #x" it will close it automatically when this is pulled. But we aren't done when this is closed, so that could lead to confusion. I guess in this case it wasn't in the form "fixes #x" but I put my defensive hat on anyway :)

File: cmd/getbackup/getbackup.go [hunk 1]
```diff
@@ -36,6 +36,10 @@ func main() {
 	// Get a list of all sites the current user has access to. Ensure we can find the site which was used in the CLI arguments in that list.
 	SiteList := &pantheon.SiteList{}
 	err := session.Request("GET", SiteList)
+	if err != nil {
+		fmt.Print("Could not complete GET request to retrieve site list.")
+		os.Exit(1)
```
Comment (rfay) [inline cmd/getbackup/getbackup.go]: wouldn't log.Fatal() work better? (And no need to use the "f" version if no args)

File: cmd/getbackup/getbackup.go [hunk 2]
```diff
@@ -45,6 +49,10 @@ func main() {
 	// Get a list of all active environments for the current site.
 	environmentList := pantheon.NewEnvironmentList(site.ID)
 	err = session.Request("GET", environmentList)
+	if err != nil {
+		fmt.Print("Could not complete GET request to retrieve environment list.")
```
Comment (rfay) [inline cmd/getbackup/getbackup.go]: I think log.Fatal() again?

File: cmd/getbackup/getbackup.go [hunk 3]
```diff
@@ -54,6 +62,10 @@ func main() {
 	// Find either a files or database backup, depending on what was asked for.
 	bl := pantheon.NewBackupList(site.ID, env.Name)
 	err = session.Request("GET", bl)
+	if err != nil {
+		fmt.Print("Could not complete GET request to retrieve backup list.")
```
Comment (beeradb) [inline cmd/getbackup/getbackup.go]: It's always a good idea to include the actual `err` output when printing messages. Good to know /why/ the request failed!

File: pkg/pantheon/site.go [hunk 4]
```diff
@@ -31,31 +31,21 @@ type Site struct {
...
+				FullName                string      `json:"full_name"`
+				GuiltyOfAbuse           interface{} `json:"guilty_of_abuse"`
+				InitialIdentityName     interface{} `json:"initial_identity_name"`
+				InitialIdentityStrategy interface{} `json:"initial_identity_strategy"`
+				LastOrgSpinup           string      `json:"last-org-spinup"`
+				Lastname                string      `json:"lastname"`
+				MinimizeJitDocs         bool        `json:"minimize_jit_docs"`
+				Modified                jsonInt64   `json:"modified"`
+				Organization            string      `json:"organization"`
+				Seens                   struct {
```
Comment (beeradb) [inline pkg/pantheon/site.go]: Pretty sure we can safely remove even more of these fields. The only ones here that I'd be inclined to keep are:

- Firstname
- FullName
- LastName
- Org

The more we can identify things that aren't being used and are unlikely to be used in the future, the less risk we take on from the inconsistencies in this API.

File: cmd/getbackup/getbackup.go (hunk 1 above)
Comment (nmccrory) [inline cmd/getbackup/getbackup.go]: I agree. log.Fatal() will work the same way and be more concise.

File: pkg/pantheon/site.go (hunk 4 above)
Comment (nmccrory) [inline pkg/pantheon/site.go]: Good to know. Let's move this discussion to a private channel and decide the final set of fields to keep.

**Comment by nmccrory** [comment]
The Circle build is currently failing because `TestSiteList` and `TestSite` are registered under the same route in the http test server. I'm still thinking out a solution to this that doesn't involve refactoring lots of other code. If anyone has suggestions, I'm all ears - enjoy the weekend 🎉

**Comment by beeradb** [comment]
@nmccrory I think you can probably skip the HTTP portion of this test and just call the unmarshal method on the site struct directly.

**Comment by nmccrory** [comment]
I resolved the HTTP registry conflict by removing the HTTP portion of `TestSite` altogether, as @beeradb suggested. The branch is now building in CircleCI successfully 💯

File: pkg/pantheon/site_test.go [hunk 5]
```diff
@@ -34,3 +34,16 @@ func TestSiteList(t *testing.T) {
 	assert.Equal(sl.Sites[0].Site.Name, "sitename1")
 	assert.Equal(sl.Sites[1].Site.Name, "sitename2")
 }
+
+// TestSite ensures Site can decode site data from the Pantheon API.
+func TestSite(t *testing.T) {
```
Comment (beeradb) [inline pkg/pantheon/site_test.go]: Please provide a more descriptive test name! happy to help you brian storm a better name if you want.

File: cmd/getbackup/getbackup.go [hunk 6]
```diff
@@ -36,6 +37,9 @@ func main() {
 	// Get a list of all sites the current user has access to. Ensure we can find the site which was used in the CLI arguments in that list.
 	SiteList := &pantheon.SiteList{}
 	err := session.Request("GET", SiteList)
+	if err != nil {
+		log.Fatalf("err: %s\nCould not complete GET request to retrieve site list.", err.Error())
```
Comment (beeradb) [inline cmd/getbackup/getbackup.go]: This is pretty minor, but you could instead do:

`log.Fatalf("err: %v\nCould not complete GET request to retrieve site list.", err)` and avoid needing to get the string value. It's the method I see most often used to print errors as strings.

File: pkg/pantheon/site_test.go [hunk 7]
```diff
@@ -34,3 +34,16 @@ func TestSiteList(t *testing.T) {
...
+	assert := assert.New(t)
+	sl := NewSiteList()
+
+	// Send JSON response back.
+	contents, err := ioutil.ReadFile("testdata/chaos_site.json")
+	assert.NoError(err)
+
+	err = sl.Unmarshal(contents)
```
Comment (beeradb) [inline pkg/pantheon/site_test.go]: It's probably a good idea to do a check on at least one field value as well. We can most likely count on `Unmarshal` to report correctly but it never hurts to make sure.

**Comment by beeradb** [comment]
🎉 We're good to go.

Thanks a ton for the work on this!

**Pull request merged**